lib/spack/spack/test/flag_handlers.py

```diff
@@ -0,0 +1,176 @@
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+import pytest
+import os
+
+import spack.spec
+import spack.build_environment
+
+
+@pytest.fixture()
+def temp_env():
+    old_env = os.environ.copy()
+    yield
+    os.environ = old_env
+
+
+def add_O3_to_cl_cflags(name, flags):
+    cl_flags = []
+    if name == 'cflags':
+        cl_flags.append('-O3')
+    return (flags, None, cl_flags)
+
+
+@pytest.mark.usefixtures('config')
+class TestFlagHandlers(object):
+    def test_no_cl_flags(self, temp_env):
+        # Test that both autotools and cmake are happy getting no cl flags
+        s1 = spack.spec.Spec('callpath')
+        s1.concretize()
+        pkg1 = spack.repo.get(s1)
+        spack.build_environment.setup_package(pkg1, False)
+
+        s2 = spack.spec.Spec('libelf')
+        s2.concretize()
+        pkg2 = spack.repo.get(s2)
+        spack.build_environment.setup_package(pkg2, False)
+
+        # Use cppflags as a canary
+        assert 'SPACK_CPPFLAGS' not in os.environ
+        assert 'CPPFLAGS' not in os.environ
+
+    def test_inject_flags(self, temp_env):
+        s = spack.spec.Spec('mpileaks cppflags=-g')
+        s.concretize()
+        pkg = spack.repo.get(s)
+        pkg.flag_handler = pkg.inject_flags
+        spack.build_environment.setup_package(pkg, False)
+
+        assert os.environ['SPACK_CPPFLAGS'] == '-g'
+        assert 'CPPFLAGS' not in os.environ
+
+    def test_env_flags(self, temp_env):
+        s = spack.spec.Spec('mpileaks cppflags=-g')
+        s.concretize()
+        pkg = spack.repo.get(s)
+        pkg.flag_handler = pkg.env_flags
+        spack.build_environment.setup_package(pkg, False)
+
+        assert os.environ['CPPFLAGS'] == '-g'
+        assert 'SPACK_CPPFLAGS' not in os.environ
+
+    def test_command_line_flags_cmake(self, temp_env):
+        s = spack.spec.Spec('callpath cppflags=-g')
+        s.concretize()
+        pkg = spack.repo.get(s)
+        pkg.flag_handler = pkg.command_line_flags
+        spack.build_environment.setup_package(pkg, False)
+
+        assert 'SPACK_CPPFLAGS' not in os.environ
+        assert 'CPPFLAGS' not in os.environ
+
+        expected = set(['-DCMAKE_C_FLAGS=-g', '-DCMAKE_CXX_FLAGS=-g',
```

Comment:
I'm confused why setting `cppflags` should also set `C_FLAGS` and `Fortran_FLAGS`

Reply:
`cppflags` is not C++ specific, it is for the C **p**re-**p**rocessor. The C pre-processor is called on C, C++, and Fortran code. I don't entirely understand why it is called on Fortran code, but I went off of the Autotools implicit make variables (https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html)